File: src/main/java/jukebot/audioutilities/AudioHandler.java
package jukebot.audioutilities;

import com.sedmelluq.discord.lavaplayer.filter.equalizer.EqualizerFactory;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioFrame;
import jukebot.JukeBot;
import jukebot.utils.Helpers;
import jukebot.utils.Permissions;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.audio.AudioSendHandler;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.TextChannel;
import net.dv8tion.jda.core.managers.AudioManager;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Random;
import java.util.concurrent.TimeUnit;

public class AudioHandler extends AudioEventAdapter implements AudioSendHandler {

    private final Permissions permissions = new Permissions();

    public AudioPlayer player;
    private EqualizerFactory equalizer = new EqualizerFactory();

    private AudioFrame lastFrame;
    private final Random selector = new Random();
    private bassBoost bassBoostMode = AudioHandler.bassBoost.OFF;

    private final LinkedList<AudioTrack> queue = new LinkedList<>();
    private final HashSet<Long> skips = new HashSet<>();
    private Long channelId;
    private Long guildId;
    private boolean shouldAnnounce = true;

    private repeatMode repeat = repeatMode.NONE;
    private boolean shuffle = false;

    private AudioTrack current = null;

    public int trackPacketLoss = 0;
    public int trackPackets = 0;

    public AudioHandler(Long guildId, AudioPlayer player) {
        this.guildId = guildId;
        this.player = player;
        player.addListener(this);
    }

    /*
     * Custom Events
     */

    boolean addToQueue(AudioTrack track, Long userID) { // boolean: shouldAnnounce
        track.setUserData(userID);

        if (!player.startTrack(track, true)) {
            queue.offer(track);
            return true;
        }

        return false;
    }

    public void stop() {
        if (isPlaying()) {
            queue.clear();
            playNext();
        }
    }

    public boolean isPlaying() {
        return player.getPlayingTrack() != null;
    }

    public boolean isBassBoosted() {
        return bassBoostMode != bassBoost.OFF;
    }

    public LinkedList<AudioTrack> getQueue() {
        return queue;
    }

    public String getRepeatMode() {
        return repeat.toString().toLowerCase();
    }

    public String getBassBoostSetting() {
        return bassBoostMode.toString().toLowerCase();
    }

    public boolean isShuffleEnabled() {
        return shuffle;
    }

    public boolean toggleShuffle() {
        return shuffle = !shuffle;
    }

    public void setRepeat(repeatMode mode) {
        repeat = mode;
    }

    public int voteSkip(Long userID) {
        skips.add(userID);
        return skips.size();
    }

    public void setChannel(Long channelId) {
        this.channelId = channelId;
    }

    public void setShouldAnnounce(Boolean shouldAnnounce) {
        this.shouldAnnounce = shouldAnnounce;
    }

    public void playNext() {
        AudioTrack nextTrack = null;

        if (repeat == repeatMode.ALL && current != null) {
            AudioTrack r = current.makeClone();
            r.setUserData(current.getUserData());
            queue.offer(r);
        }

        if (repeat == repeatMode.SINGLE && current != null) {
            nextTrack = current.makeClone();
            nextTrack.setUserData(current.getUserData());
        } else if (!queue.isEmpty()) {
            nextTrack = shuffle ? queue.remove(selector.nextInt(queue.size())) : queue.poll();
        }

        if (nextTrack != null) {
            player.startTrack(nextTrack, false);
        } else {
            current = null;
            player.stopTrack();
            bassBoost(bassBoost.OFF);

            final Guild guild = JukeBot.shardManager.getGuildById(guildId);

            if (guild == null) { // Bot was kicked or something
                JukeBot.removePlayer(guildId);
                return;
            }

            final AudioManager audioManager = guild.getAudioManager();

            if (audioManager.isConnected() || audioManager.isAttemptingToConnect()) {
                Helpers.schedule(audioManager::closeAudioConnection, 1, TimeUnit.SECONDS);

                announce("Queue Concluded!", "[Support the development of JukeBot!](https://www.patreon.com/Devoxin)");
            }
        }
    }

    private void announce(String title, String description) {
        if (!shouldAnnounce || channelId == null) {
            return;
        }

        final TextChannel channel = JukeBot.shardManager.getTextChannelById(channelId);

        if (channel == null || !permissions.canSendTo(channel)) {
            return;
        }

        channel.sendMessage(new EmbedBuilder()
                .setColor(JukeBot.embedColour)
                .setTitle(title)
                .setDescription(description)
                .build()
        ).queue(null, err -> JukeBot.LOG.error("Encountered an error while posting track announcement", err));
    }

    public void cleanup() {
        queue.clear();
        skips.clear();
        player.destroy();

        final Guild g = JukeBot.shardManager.getGuildById(guildId);
        if (g != null) {
            g.getAudioManager().setSendingHandler(null);
        }
    }

    public void bassBoost(bassBoost preset) {
        if (preset != bassBoost.OFF && bassBoostMode == bassBoost.OFF) {
            player.setFilterFactory(equalizer);
        } else if (preset == bassBoost.OFF && bassBoostMode != bassBoost.OFF) {
            player.setFilterFactory(null);
        }

        bassBoostMode = preset;
        equalizer.setGain(0, preset.getBand0());
        equalizer.setGain(1, preset.getBand1());
    }

    /*
     * Player Events
     */

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        skips.clear();
        trackPacketLoss = 0;
        trackPackets = 0;

        if (endReason.mayStartNext)
            playNext();
    }

    @Override
    public void onTrackStart(AudioPlayer player, AudioTrack track) {
        player.setPaused(false);

        if (current == null || !current.getIdentifier().equals(track.getIdentifier())) {
            current = track;
            announce("Now Playing", track.getInfo().title);
        }
    }

    @Override
    public void onTrackException(AudioPlayer player, AudioTrack track, FriendlyException exception) {
        if (repeat != repeatMode.NONE)
            repeat = repeatMode.NONE;

        announce("Playback Error", "Playback of **" + track.getInfo().title + "** encountered an error!\n" +
                exception.getLocalizedMessage());
    }

    @Override
    public void onTrackStuck(AudioPlayer player, AudioTrack track, long thresholdMs) {
        if (repeat != repeatMode.NONE)
            repeat = repeatMode.NONE;

        announce("Track Stuck", "JukeBot has automatically detected a stuck track and will now play the next song in the queue.");
        playNext();
    }

    /*
     * Packet Sending Events
     */

    @Override
    public boolean canProvide() {
        lastFrame = player.provide();

        if (!player.isPaused()) {
            if (lastFrame == null) {
                trackPacketLoss++;
            } else {
                trackPackets++;
            }
        }

        return lastFrame != null;
    }

    @Override
    public byte[] provide20MsAudio() {
        return lastFrame.getData();
    }

    @Override
    public boolean isOpus() {
        return true;
    }

    public enum repeatMode {
        SINGLE, ALL, NONE
    }

    public enum bassBoost {
        OFF(0F, 0F),
        LOW(0.25F, 0.15F),
        MEDIUM(0.50F, 0.25F),
        HIGH(0.75F, 0.50F),
        INSANE(1F, 0.75F);

        private final float band0;
        private final float band1;

        bassBoost(float b0, float b1) {
            this.band0 = b0;
            this.band1 = b1;
        }

        public float getBand0() {
            return band0;
        }

        public float getBand1() {
            return band1;
        }
    }

}

File: src/main/kotlin/me/devoxin/jukebot/audio/AudioHandler.kt
package me.devoxin.jukebot.audio

import com.sedmelluq.discord.lavaplayer.format.StandardAudioDataFormats
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import com.sedmelluq.discord.lavaplayer.track.playback.AudioFrameFlags
import com.sedmelluq.discord.lavaplayer.track.playback.MutableAudioFrame
import io.sentry.Sentry
import io.sentry.event.BreadcrumbBuilder
import io.sentry.event.Event
import io.sentry.event.EventBuilder
import io.sentry.event.interfaces.ExceptionInterface
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import me.devoxin.flight.internal.utils.TextUtils
import me.devoxin.jukebot.Database
import me.devoxin.jukebot.Launcher
import me.devoxin.jukebot.audio.AudioHandler.RepeatMode.ALL
import me.devoxin.jukebot.audio.AudioHandler.RepeatMode.SINGLE
import me.devoxin.jukebot.audio.sources.spotify.SpotifyAudioTrack
import me.devoxin.jukebot.extensions.await
import me.devoxin.jukebot.extensions.capitalise
import me.devoxin.jukebot.extensions.toTimeString
import me.devoxin.jukebot.extensions.truncate
import me.devoxin.jukebot.utils.Components
import me.devoxin.jukebot.utils.Helpers
import me.devoxin.jukebot.utils.Scopes
import me.devoxin.jukebot.utils.collections.FixedDeque
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission.*
import net.dv8tion.jda.api.audio.AudioSendHandler
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.utils.messages.MessageCreateBuilder
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.TimeUnit

class AudioHandler(private val guildId: Long,
                   var channelId: Long,
                   private val initialVoiceChannelId: Long,
                   val player: AudioPlayer) : AudioEventAdapter(), AudioSendHandler {
    private val mutex = Mutex()

    private val mutableFrame = MutableAudioFrame()
    private val buffer = ByteBuffer.allocate(StandardAudioDataFormats.DISCORD_OPUS.maximumChunkSize())

    private var initialConnect = false

    // Playback Settings
    val bassBooster = BassBooster(player)
    var repeat = RepeatMode.NONE
    var shuffle = false

    val queue = LinkedList<AudioTrack>()
    private val skips = hashSetOf<Long>()

    var lastAnnouncement: Message? = null
    var shouldAnnounce = true

    // Performance Tracking
    var trackPacketLost = 0
        private set
    var trackPacketsSent = 0
        private set

    // Player Stuff
    private val autoPlay = AutoPlay(guildId)

    private val history = FixedDeque<AudioTrack>(10)
    private var backwards = false
    val canGoBack: Boolean get() = history.isNotEmpty()

    val isPlaying: Boolean get() = player.playingTrack != null
    val isEncoding: Boolean get() = AudioFrameFlags.OPUS_PASSTHROUGH !in mutableFrame.flags

    private val guild: Guild?
        get() = Launcher.shardManager.getGuildById(guildId)

    init {
        player.addListener(this)
        mutableFrame.setBuffer(buffer)
    }

    /**
     * Enqueues a track to be played.
     * @param track
     *        The track to add to the queue, or to begin playing.
     * @param userId
     *        The ID of the user who requested the track.
     *
     * @return true, if the calling function should announce that the track has been enqueued.
     */
    fun enqueue(track: AudioTrack, userId: Long, playNext: Boolean): Boolean {
        if (!initialConnect) {
            val guild = guild
            val channel = guild?.getVoiceChannelById(initialVoiceChannelId)

            if (guild == null || channel == null) {
                Launcher.playerManager.removePlayer(guildId)
                return false
            }

            guild.audioManager.openAudioConnection(channel)
            initialConnect = true
        }

        track.userData = userId

        if (!startTrack(track, true)) {
            if (playNext) {
                queue.add(0, track)
            } else {
                queue.add(track)
            }
            return true
        }

        return false
    }

    fun voteSkip(userID: Long): Int {
        skips.add(userID)
        return skips.size
    }

    fun previous() {
        if (!canGoBack) {
            return
        }

        val track = player.playingTrack
        backwards = true

        if (track != null) {
            queue.add(0, track.makeClone())
        }

        startTrack(history.removeLast().makeClone())
    }

    fun next(shouldAutoPlay: Boolean = true, lastTrack: AudioTrack? = player.playingTrack) {
        var nextTrack: AudioTrack? = null

        if (lastTrack is SpotifyAudioTrack && lastTrack.position >= minOf(lastTrack.duration * 0.20, 30000.0)) {
            autoPlay.store(lastTrack)
        }

        if (lastTrack != null && repeat != RepeatMode.NONE) {
            val cloned = lastTrack.makeClone()

            when (repeat) {
                ALL -> queue.offer(cloned)
                SINGLE -> nextTrack = cloned
                else -> {}
            }
        }

        if (nextTrack == null && !queue.isEmpty()) {
            nextTrack = if (shuffle) queue.removeAt(selector.nextInt(queue.size)) else queue.poll()
        }

        if (nextTrack != null) {
            startTrack(nextTrack)
            return
        }

        if (shouldAutoPlay && autoPlay.enabled && autoPlay.isUsable) {
            val recommendedTrack = autoPlay.getRelatedTrack()

            if (recommendedTrack != null) {
                startTrack(recommendedTrack)
                return
            }
            
            announce("AutoPlay", "AutoPlay was unable to find a track to play.", set = false)
        }

        announce("Queue Concluded", "Keep the party going by adding more tracks!", set = false)
        player.stopTrack()
        bassBooster.boost(0)

        val audioManager = guild?.audioManager
            ?: return Launcher.playerManager.removePlayer(guildId).let { cleanup() }

        if (audioManager.isConnected) {
            Helpers.schedule(audioManager::closeAudioConnection, 1, TimeUnit.SECONDS)
        }

        cleanup()
    }

    private fun announce(title: String?,
                         description: String,
                         thumbnailUrl: String? = null,
                         set: Boolean = true,
                         extra: MessageCreateBuilder.() -> Unit = {}) {
        if (!shouldAnnounce) {
            return
        }

        val channel = guild?.getChannelById(TextChannel::class.java, channelId)
            ?.takeIf { it.guild.selfMember.hasPermission(VIEW_CHANNEL, MESSAGE_SEND, MESSAGE_EMBED_LINKS) }
            ?: return

        Scopes.IO.launch {
            announce0(channel, title, description, thumbnailUrl, set, extra)
        }
    }

    private suspend fun announce0(channel: GuildMessageChannel,
                                  title: String?,
                                  description: String,
                                  thumbnailUrl: String?,
                                  set: Boolean,
                                  extra: MessageCreateBuilder.() -> Unit) {
        mutex.withLock {
            lastAnnouncement?.runCatching { delete().await() }

            channel.runCatching {
                sendMessage(MessageCreateBuilder()
                    .addEmbeds(EmbedBuilder()
                        .setColor(Database.getColour(guildId))
                        .setTitle(title)
                        .setDescription(description.truncate(1000))
                        .setThumbnail(thumbnailUrl)
                        .build())
                    .apply(extra)
                    .build())
                    .await()
            }.onSuccess {
                if (set) lastAnnouncement = it
            }
        }
    }

    private fun setNick(nick: String?) {
        guild?.selfMember
            ?.takeIf { it.hasPermission(NICKNAME_CHANGE) && Database.getIsPremiumServer(guildId) && Database.getIsMusicNickEnabled(guildId) }
            ?.modifyNickname(nick)?.queue()
    }

    fun cleanup() {
        queue.clear()
        skips.clear()
        player.destroy()

        guild?.audioManager?.sendingHandler = null
        setNick(null)
        lastAnnouncement?.runCatching { delete().queue() }
    }

    private fun startTrack(track: AudioTrack, noInterrupt: Boolean = false): Boolean {
        if (track is HighQualityAudioTrack) {
            track.setAllowHighQuality(Database.getIsPremiumServer(guildId))
        }

        return player.startTrack(track, noInterrupt)
    }

    /*
     *  +===================+
     *  |   Player Events   |
     *  +===================+
     */
    override fun onTrackStart(player: AudioPlayer, track: AudioTrack) {
        player.isPaused = false

        if (history.isNotEmpty() && track.identifier == history.first().identifier) {
            // TODO check if last sent message ID is the now playing message?
            return
        }

        val duration = track.takeIf { !it.info.isStream }?.duration?.toTimeString() ?: "∞"
        val requester = if (track.userData as Long == Launcher.shardManager.botId) "AutoPlay" else "<@${track.userData}>"

        announce(null, "**${track.info.title}**\n*${track.info.author} — $duration*\n$requester", (track as? SpotifyAudioTrack)?.artworkUrl) {
            setComponents(Components.nowPlayingRowUnpaused)
        }

        val nick = TextUtils.truncate("${track.info.title} - ${track.info.author}", 32)
        setNick(nick)
    }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        skips.clear()
        trackPacketLost = 0
        trackPacketsSent = 0

        if (!backwards) {
            history.add(track)
        }

        backwards = false

        if (endReason.mayStartNext) {
            next(lastTrack = track)
        }
    }

    override fun onTrackException(player: AudioPlayer, track: AudioTrack, exception: FriendlyException) {
        val breadCrumb = BreadcrumbBuilder()
            .setCategory("AudioHandler")
            .setMessage("Track ID: ${track.identifier}")
            .build()

        val eventBuilder = EventBuilder()
            .withMessage(exception.message)
            .withLevel(Event.Level.ERROR)
            .withSentryInterface(ExceptionInterface(exception))
            .withBreadcrumbs(listOf(breadCrumb))

        Sentry.capture(eventBuilder)
        repeat = RepeatMode.NONE

        announce("Track Unavailable", "An error occurred during the playback of **${track.info.title}**\nSkipping...", set = false)
    }

    override fun onTrackStuck(player: AudioPlayer, track: AudioTrack, thresholdMs: Long) {
        repeat = RepeatMode.NONE

        announce("Track Unavailable", "The track **${track.info.title}** has frozen and cannot be played\nSkipping...", set = false)
        next(lastTrack = track)
    }

    /*
     *  +=======================+
     *  |   JDA Audio Sending   |
     *  +=======================+
     */
    override fun canProvide(): Boolean {
        val frameProvided = player.provide(mutableFrame)

        if (!player.isPaused) {
            if (!frameProvided) {
                trackPacketLost++
            } else {
                trackPacketsSent++
            }
        }

        return frameProvided
    }

    override fun provide20MsAudio(): ByteBuffer = buffer.flip()
    override fun isOpus() = true

    enum class RepeatMode {
        ALL,
        SINGLE,
        NONE;

        fun humanized(): String {
            return this.toString().lowercase().capitalise()
        }
    }

    companion object {
        const val EXPECTED_PACKET_COUNT_PER_MIN = ((60 * 1000) / 20).toDouble()
        private val selector = Random()
    }
}
